File: 2023.1/labs/word_count/go/word_count.go
package main

import (
    "fmt"
    "io/ioutil"
    "log"
    "os"
    "path/filepath"
    "strings"
    "sync"
)

func wc(fileContent string) int {
    words := strings.Fields(fileContent)
    return len(words)
}

func wc_file(filePath string) int {
    fileContent, err := ioutil.ReadFile(filePath)
    if err != nil {
        log.Fatal(err)
    }

    return wc(string(fileContent))
}

func wc_dir(directoryPath string) int {
    files, err := ioutil.ReadDir(directoryPath)
    if err != nil {
        log.Fatal(err)
    }

    var numberOfWords int
    for _, file := range files {
        if file.IsDir() {
            continue
        }

        filePath := filepath.Join(directoryPath, file.Name())
        numberOfWords += wc_file(filePath)
    }

    return numberOfWords
}

func main() {
    if len(os.Args) < 2 {
        fmt.Println("Usage: go run word_count.go <directory>")
        os.Exit(1)
    }

    directoryPath := os.Args[1]
    var numberOfWords int

    // Get all subdirectories in the given directory.
    subdirs, err := ioutil.ReadDir(directoryPath)
    if err != nil {
        log.Fatal(err)
    }

    // Create a WaitGroup to wait for all goroutines to finish.
    var wg sync.WaitGroup

    // Iterate over each subdirectory and start a goroutine to count the words.
    for _, subdir := range subdirs {
        if !subdir.IsDir() {
            continue
        }

        subdirPath := filepath.Join(directoryPath, subdir.Name())
        wg.Add(1)

        go func(subdirPath string) {
            defer wg.Done()
            numberOfWords += wc_dir(subdirPath)
        }(subdirPath)
    }

    // Wait for all goroutines to finish.
    wg.Wait()

    fmt.Println(numberOfWords)
}

File: 2023.1/labs/word_count/java/src/main/java/WordCount.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.ReentrantLock;

public class WordCount {
    
    private static ReentrantLock lock = new ReentrantLock();
    private static int count = 0; 
    
    public static void main(String[] args) {
        if (args.length != 1) {
            System.err.println("Usage: java WordCount <root_directory>");
            System.exit(1);
        }

        String rootPath = args[0];
        File rootDir = new File(rootPath);
        File[] subdirs = rootDir.listFiles();

        if (subdirs != null) {
	    
     	    List<Thread> threads = new ArrayList<>();
	    
	    for (File subdir : subdirs) {
                if (subdir.isDirectory()) {
                    File[] files = subdir.listFiles();
                    if (files != null) {
                        for (File file : files) {
                            if (file.isFile()) {
                                Thread thread = new Thread(new MyThread(file.getAbsolutePath()));
                                threads.add(thread);
                                thread.start();
                            }
                        }
                    }
                }
            }

	    for (Thread thread : threads) {
                try {
                    thread.join();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }

        System.out.println(count);
    }

    private static class MyThread implements Runnable {
        private String filePath;

	MyThread(String filePath){
	    this.filePath = filePath;
	}

	@Override
	public void run(){
	    int localCount = wcFile(filePath);
	    lock.lock();
	    try {
	    	count += localCount;
	    } finally {
	    	lock.unlock();
	    }
	}
    }

    public static int wc(String fileContent) {
        String[] words = fileContent.split("\\s+");
        return words.length;
    }

    public static int wcFile(String filePath) {
        try {
            BufferedReader reader = new BufferedReader(new FileReader(filePath));
            StringBuilder fileContent = new StringBuilder();
            String line;

            while ((line = reader.readLine()) != null) {
                fileContent.append(line).append("\n");
            }

            reader.close();
            return wc(fileContent.toString());

        } catch (IOException e) {
            e.printStackTrace();
            return -1;
        }
    }
}
